KEP-32: Community Repository Management (#1574)
kudobuilder/kudo: keps/0032-community-repository-management.md

### Goals

- Provide a simple workflow for upstream operator packages to get added to the community repository
- Remove the need to "host" copies of upstream operator packages in a Git repository
- Provide a mechanism to define upstream operator package sources
- Provide a mechanism to define upstream operator package maturity

### Non-Goals

- Change the practice of creating a PR against a Git repository to add operator packages to the community repository
- Change operator package management as defined in [KEP-10](0010-package-manager.md) and [KEP-15](0015-repository-management.md)
- Integration with [ArtifactHub](https://artifacthub.io/)

## Proposal

> Upstream operator developers still create PRs against a Git repository to add their operator packages to the community repository. This Git repository lists references to upstream operator packages instead of a full copy of an operator package. A reference can point to a Git resitory or a package tarball. A version of an operator package is described by a specific tag of a Git repository or a URL pointing to an operator tarball of that release.

Reviewer: ```suggestion
Upstream operator developers still create PRs against a Git repository to add their operator packages to the community repository. This Git repository lists references to upstream operator packages instead of a full copy of an operator package. A reference can point to a Git repository or a package tarball. A version of an operator package is described by a specific tag of a Git repository or a URL pointing to an operator tarball of that release.
```

> For example, consider an operator package developed at `github.com/example/example-operator` that has  tagged versions `1.0.0`, `1.1.0` and the operator package in the `operator` folder. There's also a tarball for version `0.9.0` hosted at `example.org/example-operator-0.9.0.tgz`. To add or update this operator package, the developers would create a PR referencing their upstream Git repository and the specific version, e.g. by adding a file `example-operator.yaml` like

Reviewer: ```suggestion
For example, consider an operator package developed at `github.com/example/example-operator` that has tagged versions `1.0.0`, `1.1.0` and the operator package in the `operator` folder. There's also a tarball for version `0.9.0` hosted at `example.org/example-operator-0.9.0.tgz`. To add or update this operator package, the developers would create a PR referencing their upstream Git repository and the specific version, e.g. by adding a file `example-operator.yaml` like
```

```yaml

> api: v1alpha1

Reviewer 2: we should include the domain here. `operators.kudo.dev/v1alpha1` maybe?

Author: You're right. Didn't update the KEP yet, doing this already in `kitt`, using `index.kudo.dev/v1alpha1` there.

name: Example Operator
git-sources:
- name: git-repo
  url: github.com/example/example-operator.git
versions:
- version: "0.9.0"
  url: example.org/example-operator-0.9.0.tgz
- version: "1.0.0"
  git:
    source: git-repo
    tag: "1.0.0"
    dir: operator
- version: "1.1.0"
  git:
    source: git-repo
    tag: "1.1.0"
    dir: operator
```

Once this PR is merged, CI tooling detects the new YAML file, clones the referenced upstream Git repository, checks out the tag, and adds the operator package in the specified folder to the existing index. This workflow is similar to [krew-index](https://github.com/kubernetes-sigs/krew-index). Of course, CI tests that don't update the community repository can run before the PR is merged. These tests include checking the referenced operator package for validity. Additional conformance testing can be added as well.

We can add more metadata to the YAML reference file. E.g., support for different upstream sources like Mercurial.

> The maturity level of an operator package can be provided in this file as well, though this should be added to the operator package instead.

Reviewer: does the maturity level of an operator is just a metadata field? 
I would really like to see some defined approach here with testing which can prove that an operator can be upgraded for example. 

What is making KUDO Kafka `1.3.1` as a better operator vs version `0.1.0`?
If we have defined criteria which we agree upon, we could achieve non-biased maturity levels

---
kep-number: 32
title: Community Repository Management
short-desc: Details on how to add operator to the community repository
authors:
  - "@nfnt"
owners:
  - "@nfnt"
creation-date: 2020-07-02
last-updated: 2020-07-02
status: provisional
see-also:
  - KEP-10
  - KEP-15
---

# Community Repository Management

## Table of Contents

- [Community Repository Management](#community-repository-management)
  - [Table of Contents](#table-of-contents)
  - [Summary](#summary)
  - [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals](#non-goals)
  - [Proposal](#proposal)
    - [Risks and Mitigations](#risks-and-mitigations)
  - [Graduation Criteria](#graduation-criteria)
  - [Implementation History](#implementation-history)

## Summary

> By default, every KUDO deployment installs operator packages from the community repository. As we encourage the KUDO community to add new operator packages to this repository, this KEP defines

Reviewer 3: ```suggestion
By default, every KUDO deployment installs operator packages from the community repository. As we encourage the KUDO community to add new operator packages to this repository, this KEP defines:
```

- The workflow to add new operator packages to the community repository
- Rules and means to describe the maturity of operator packages in the community repository

## Motivation

> Currently, every operator package in the community repository is provided from the `kudobuilder/operators` Git repository. This Git repository contains package definitions, documentation and sometimes tests for each operator. While it is convenient, to have this all in a single repository, this is challenging for larger or third-party operators:

Reviewer 3: I'm not sure how important this is... but this isn't entirely true.   The package repo (google bucket) is manually managed and the `kudobuilder/operators` is the main source of truth... however we have added operators that are no in that repository.    An example is "galera".  

```
k kudo search gal
Name  	Operator Version	App Version
galera	0.1.0           	10.5.3     
```

Reviewer 3: this is not true:  `Currently, every operator package in the community repository is provided from the `kudobuilder/operators` Git repository.`

Author: I'll rephrase that sentence.

> - Large operators usually provide own Docker images and/or additional dependencies. These have to be build, tested and deployed. The tools for this can't be provided by `kudobuilder/operators`.

Reviewer 4: ```suggestion
- Large operators usually provide their own Docker images and/or additional dependencies. These have to be built, tested and deployed. The tools for this can't be provided by `kudobuilder/operators`.
```

- Some operator tests cannot be covered by the test tooling provided in `kudobuilder/operators` making it necessary to host the tests in a separate Git repository.
- Providing multiple versions of an operator is only possible by using separate folders. Teams developing operators might prefer different workflows (e.g. Git tags)

These requirements make it necessary for operator packages to be developed in a separate Git repository. As a result, the operator packages in `kudobuilder/operators` are copies of specific versions of the respective upstream Git repository. This approach has challenges as well, because bugs discovered in an operator need to be resolved in the upstream Git repository, not in `kudobuilder/operators`. I.e., there needs to be metadata to link to the upstream Git repository. By having some form of metadata describing upstream operator package sources, the same metadata can be used to describe other properties of a package, e.g. the maturity level.

> ### Goals

Reviewer 4: Should tackling how we integrate with Aritfact Hub be a goal or non-goal:

https://github.com/artifacthub/hub

Author: That should be a non-goal for now.

- Provide a simple workflow for upstream operator packages to get added to the community repository
- Remove the need to "host" copies of upstream operator packages in a Git repository
- Provide a mechanism to define upstream operator package sources
- Provide a mechanism to define upstream operator package maturity

### Non-Goals

- Change the practice of creating a PR against a Git repository to add operator packages to the community repository
- Change operator package management as defined in [KEP-10](0010-package-manager.md) and [KEP-15](0015-repository-management.md)

## Proposal

> Upstream operator developers still create PRs against a Git repository to add their operator packages to the community repository. This Git repository lists references to upstream operator packages instead of a full copy of an operator package. For example, consider an operator package developed at `github.com/example/example-operator` that has a tagged version `1.0.0` and the operator package in the `operator` folder. To add or update this operator package, the developers would create a PR referencing their upstream Git repository and the specific version, e.g. by adding a file `example-operator.yaml` like

Author: Depends on the folder structure we choose. If we follow the example of krew, we would have one folder in which all operator reference files live. In that case it's beneficial to name these files like their operator. E.g.
```
operators/cassandra.yaml
operators/kafka.yaml
operators/mysql.yaml
```
We could also choose a model where there is a folder per operator. In that case `upstream.yaml` makes sense, e.g.
```
operators/cassandra/upstream.yaml
operators/kafka/upstream.yaml
operators/mysql/upstream.yaml
```
Conceptually there isn't any difference for the implementation. Though, having a folder per operator would make it possible to add additional files related to that operator in the future. Can't think of any real use-cases for that though. Hmm, maybe something like `icon.png` that could be used once we have a web UI for the community repo maybe?

> Upstream operator developers still create PRs against a Git repository to add their operator packages to the community repository. This Git repository lists references to upstream operator packages instead of a full copy of an operator package. For example, consider an operator package developed at `github.com/example/example-operator` that has a tagged version `1.0.0` and the operator package in the `operator` folder. To add or update this operator package, the developers would create a PR referencing their upstream Git repository and the specific version, e.g. by adding a file `example-operator.yaml` like

Reviewer 3: I like the krew alignment
I'm not a fan of file in the upstream repo... there needs to be checks on the release of artifacts to the bucket.

```yaml

> name: Example Operator

Author: Yes, needs to be versioned, good point. I also agree on the maintainer section, though there's already a maintainer section in the actual operator. Not sure if there's much value in duplicating it here.

> name: Example Operator

Reviewer: +1 to versioned and maintainers
Maintainers for the operator itself and responsible for testing and making sure it makes it to community repo could be a different set of users

versions:
- version: "1.0.0"
  git:
    repository: github.com/example/example-operator.git
    tag: "1.0.0"
    dir: operator
```

> 

Reviewer 4: Should we be responsible for compiling the operator as part of the KUDO community repo, or should we be referencing releases that are built and hosted on GitHub (e.g. like Krew does https://github.com/kubernetes-sigs/krew-index/blob/master/plugins/kudo.yaml)?

By referencing a release, we could eventually support various release objects (e.g. helm charts/Docker Images/OCI Artifact Specs/etc) rather than requiring the internal objects to build those objects.

Author: It depends on what "compiling" should cover. We should be responsible for creating tarballs from operator package folders because these tarballs are specific to repositories and operator developers shouldn't keep them as part of their Git repositories. Though we could also support these tarballs directly if developers want to do that.

Reviewer 5: Agreed. Creating the tarballs ourself also makes sure that they're not changed afterwards

Reviewer 3: I'm ok with having a model that creates the tarballs... but I also want to support referencing prebuilt tars.   Requiring us to build them assumes that the build process is exactly how we define it... and that the git repo has an instance of this...  I could imagine additional templates or build techniques for which the files to create the operator are an intermediary step (which an org may not want to check-in, outside of our requirement on them).   Supporting prebuilts is necessary IMO.   We should: 1) either make this a current non-goal (for future definition) or 2) define it.

Reviewer 3: The use of "version" concerns me a bit... 

context:  The version of an operator is defined in the operator.yaml and is part of the tarball name.   The tarball version also defines this as app_version and op_version.   so... 

1. Is the "version" used here... app, or op?
2. Is this expected to be the app_V-op_v?
3. Is it checked / verified against the artifact?  does the PR fail if it doesn't match?

Additionally... can we make it easy for people... can we assume that the "tag" is the "version" if not specified?  It seems like we could assume that the version is the tag but allow tag to override that assumption.

Author: Each package that is to be added to the community repo needs to be checked for _validity_ by the CI. This would at least cover that it is recognized by KUDO as an operator package. More checks for conformance are of course possible as well. All of these checks will work with directories as well as tarballs. When referencing tarballs, these will be copied into the community repository. I'll clarify this in the KEP.

@kensipe Good point regarding the version. This is again a question on which metadata should in these reference files versus what metadata is provided by the actual operator package. A reference file could be as simple as
```
versions:
-   git:
    repository: github.com/example/example-operator.git
    tag: "1.0.0"
    dir: operator
```
because the other metadata (`name` as well as `version`) is redundant. But IMO there's value to have them duplicated here, as it helps developers to understand what is being referenced. These values will show up in CI logs and could help debug CI failures, e.g. when using a wrong Git tag.

> Once this PR is merged, CI tooling detects the new YAML file, clones the referenced upstream Git repository, checks out the tag, and adds the operator package in the specified folder to the existing index. Of course, CI tests that don't update the community repository can run before the PR is merged. This workflow is similar to [krew-index](https://github.com/kubernetes-sigs/krew-index)

Author: I left testing out, this needs to be discussed. IMO, upstream operators should be responsible for testing. The community repo tooling should only be concerned with testing for conformance which should run tests that work for every operator.

> Once this PR is merged, CI tooling detects the new YAML file, clones the referenced upstream Git repository, checks out the tag, and adds the operator package in the specified folder to the existing index. Of course, CI tests that don't update the community repository can run before the PR is merged. This workflow is similar to [krew-index](https://github.com/kubernetes-sigs/krew-index)

Reviewer 3: Agree that the larger conversation around testing should be discussed and implemented... and that we don't need to include it here (for now).  But shouldn't we cover:
1. failures in access to bucket (network or auth)
2. does the PR merge if the failure occurred?  IS there a notification if this failure occurs?
3. Is there a check on version?  is that possible to fail? and what happens? / how is it resolved?
4. What is github is down?  where the git clone fails?  or the repository is deleted?

> Once this PR is merged, CI tooling detects the new YAML file, clones the referenced upstream Git repository, checks out the tag, and adds the operator package in the specified folder to the existing index. Of course, CI tests that don't update the community repository can run before the PR is merged. This workflow is similar to [krew-index](https://github.com/kubernetes-sigs/krew-index)

Author: Aren't these points implementation details? Not sure if they should be covered here. IMO, we need to distinguish between conformance tests that are independent of an operator -- e.g., that the referenced Git repo can be cloned, the referenced operator is valid, ... -- and tests specific to an operator -- e.g. a `kuttl-test.yaml` specified by metadata.

> Once this PR is merged, CI tooling detects the new YAML file, clones the referenced upstream Git repository, checks out the tag, and adds the operator package in the specified folder to the existing index. Of course, CI tests that don't update the community repository can run before the PR is merged. This workflow is similar to [krew-index](https://github.com/kubernetes-sigs/krew-index)

Reviewer: I would really like to see tests here. 
We should ask for minimal tests with a timeout of 5 min and we could use them to perform sanity before merging PRs

For example, a default version of an operator might fail due to prometheus operator requirement, so doing simple installation might not be an answer here. But we shouldn't explore and discover how to do a minimal installation that works, that should be provided by operator developer. And tests that consume a limited amount of resources and with a reasonable timeout would help us improve a bit the quality here.

> We can add more metadata to the YAML reference file. E.g., support for different upstream sources like Mercurial.

Reviewer 2: Should we want to generate a web presence for the operators (e.g. like Helm Hub) we'd need more info e.g. maintainer info (URL and whatnot) and operator description (README, installation instructions, list of params, that sort of things). Should this be part of the upstream repo?

Author: To be discussed. Ideally, all of this information should be provided as part of the operator package.

The maturity level of an operator package can be provided in this file as well.

> Other useful metadata would be links to the operator homepage or documentation.

Reviewer 4: How would these differ from the internal values to the Operator/OperatorVersion these contain.  e.g. https://github.com/kudobuilder/kudo/blob/main/pkg/apis/kudo/v1beta1/operator_types.go#L28

Author: We need to discuss which values make sense here versus having them in the operator package. IMO, most metadata should be provided by the operator package.

Reviewer 3: Yeah.. I prefer metadata to be in the the tarball to be retrieved if needed by this tool.  There are things that likely make sense to have outside the bundle... still thinking

Author: Agreed. As mentioned [here](https://github.com/kudobuilder/kudo/pull/1574/files#r449450608), it can make sense to duplicate some metadata though, as this can help debug CI failures.
